File: src/SheetReader.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.IO.Compression;
using System.Text;
using System.Xml;

namespace PaymentAlert
{
    /// <summary>표 한 장. 엑셀은 시트 하나, CSV 는 파일 하나.</summary>
    public sealed class Sheet
    {
        public string 이름 = "";
        public readonly List<string[]> 행 = new List<string[]>();
    }

    /// <summary>
    /// ERP 에서 받은 엑셀(.xlsx)·CSV 를 글자 표로 읽는다 (ADR-0023).
    /// .NET Framework 에 들어 있는 압축·XML 기능만 쓴다 — 추가 DLL 없음.
    /// 숫자는 엑셀에 저장된 그대로의 글자(예: 날짜 일련번호 46269)로 돌려주고, 뜻은 읽는 쪽이 정한다.
    /// </summary>
    public static class SheetReader
    {
        public const long 최대크기 = 20L * 1024 * 1024;
        const long 풀린최대 = 64L * 1024 * 1024;   // 압축 폭탄 막기: 한 부분이 풀렸을 때의 한도
        const int 최대행 = 20000;
        const int 최대열 = 200;

        public sealed class 읽기오류 : Exception
        {
            public 읽기오류(string message) : base(message) { }
        }

        public static List<Sheet> Read(byte[] data, string fileName)
        {
            if (data == null || data.Length == 0) throw new 읽기오류("빈 파일입니다.");
            if (data.Length > 최대크기) throw new 읽기오류("20MB 보다 큰 파일은 읽지 않습니다.");

            if (data.Length >= 4 && data[0] == 0x50 && data[1] == 0x4B && data[2] == 0x03 && data[3] == 0x04)
                return Xlsx(data);

            // D0 CF 11 E0: 옛 엑셀(.xls) 또는 암호·문서보안이 걸린 엑셀. 둘 다 내용을 읽을 수 없다.
            if (data.Length >= 4 && data[0] == 0xD0 && data[1] == 0xCF && data[2] == 0x11 && data[3] == 0xE0)
                throw new 읽기오류("암호가 걸렸거나 문서보안(DRM)이 적용된 파일, 또는 옛 엑셀(.xls) 파일이라 읽을 수 없습니다. " +
                                  "ERP 에서 CSV 로 받거나, 엑셀에서 암호·보안을 푼 뒤 .xlsx 로 저장해 올려 주세요.");

            string ext = (Path.GetExtension(fileName ?? "") ?? "").ToLowerInvariant();
            if (ext == ".xlsx" || ext == ".xlsm")
                throw new 읽기오류("엑셀 파일 형식이 아닙니다. 파일이 손상됐거나 확장자만 바뀐 파일입니다.");

            var sheet = Csv(Text(data));
            sheet.이름 = Path.GetFileNameWithoutExtension(fileName ?? "") ?? "";
            return new List<Sheet> { sheet };
        }

        // ── CSV ─────────────────────────────────────────────────

        /// <summary>UTF-8(BOM 유무) → 안 되면 한국어 윈도 코드(CP949). ERP 의 CSV 는 대개 CP949 다.</summary>
        static string Text(byte[] data)
        {
            if (data.Length >= 3 && data[0] == 0xEF && data[1] == 0xBB && data[2] == 0xBF)
                return new UTF8Encoding(false).GetString(data, 3, data.Length - 3);
            try
            {
                return new UTF8Encoding(false, true).GetString(data);
            }
            catch (DecoderFallbackException)
            {
                return Encoding.GetEncoding(949).GetString(data);
            }
        }

        public static Sheet Csv(string text)
        {
            var sheet = new Sheet();
            if (string.IsNullOrEmpty(text)) return sheet;

            // 첫 줄에 쉼표가 없고 탭이 있으면 탭 구분으로 본다 (엑셀 '텍스트(탭 구분)' 저장).
            int eol = text.IndexOfAny(new[] { '\r', '\n' });
            string first = eol < 0 ? text : text.Substring(0, eol);
            char sep = first.IndexOf(',') < 0 && first.IndexOf('\t') >= 0 ? '\t' : ',';

            var row = new List<string>();
            var cell = new StringBuilder();
            bool quoted = false;
            for (int i = 0; i < text.Length; i++)
            {
                char ch = text[i];
                if (quoted)
                {
                    if (ch == '"')
                    {
                        if (i + 1 < text.Length && text[i + 1] == '"') { cell.Append('"'); i++; }
                        else quoted = false;
                    }
                    else cell.Append(ch);
                    continue;
                }
                if (ch == '"' && cell.Length == 0) { quoted = true; continue; }
                if (ch == sep) { row.Add(cell.ToString()); cell.Length = 0; continue; }
                if (ch == '\r' || ch == '\n')
                {
                    if (ch == '\r' && i + 1 < text.Length && text[i + 1] == '\n') i++;
                    row.Add(cell.ToString()); cell.Length = 0;
                    행추가(sheet, row);
                    row = new List<string>();
                    continue;
                }
                cell.Append(ch);
            }
            if (cell.Length > 0 || row.Count > 0)
            {
                row.Add(cell.ToString());
                행추가(sheet, row);
            }
            return sheet;
        }

        static void 행추가(Sheet sheet, List<string> row)
        {
            if (sheet.행.Count >= 최대행) throw new 읽기오류("행이 너무 많습니다 (" + 최대행 + "행까지).");
            if (row.Count > 최대열) row.RemoveRange(최대열, row.Count - 최대열);
            sheet.행.Add(row.ToArray());
        }

        // ── XLSX ────────────────────────────────────────────────

        const string NsMain = "http://schemas.openxmlformats.org/spreadsheetml/2006/main";
        const string NsRel = "http://schemas.openxmlformats.org/officeDocument/2006/relationships";
        const string NsPkgRel = "http://schemas.openxmlformats.org/package/2006/relationships";

        static List<Sheet> Xlsx(byte[] data)
        {
            try
            {
                using (var ms = new MemoryStream(data, false))
                using (var zip = new ZipArchive(ms, ZipArchiveMode.Read))
                {
                    XmlDocument wb = Part(zip, "xl/workbook.xml");
                    if (wb == null) throw new 읽기오류("엑셀 통합 문서 정보(workbook.xml)가 없습니다.");

                    var targets = new Dictionary<string, string>(StringComparer.Ordinal);
                    XmlDocument rels = Part(zip, "xl/_rels/workbook.xml.rels");
                    if (rels != null)
                    {
                        foreach (XmlElement r in rels.GetElementsByTagName("Relationship", NsPkgRel))
                        {
                            string t = r.GetAttribute("Target");
                            if (t.StartsWith("/")) t = t.TrimStart('/');
                            else t = "xl/" + t;
                            targets[r.GetAttribute("Id")] = t;
                        }
                    }

                    List<string> shared = SharedStrings(zip);
                    var result = new List<Sheet>();
                    foreach (XmlElement s in wb.GetElementsByTagName("sheet", NsMain))
                    {
                        string rid = s.GetAttribute("id", NsRel);
                        string path;
                        if (!targets.TryGetValue(rid, out path)) continue;
                        XmlDocument doc = Part(zip, path);
                        if (doc == null) continue;
                        Sheet sheet = SheetOf(doc, shared);
                        sheet.이름 = s.GetAttribute("name");
                        result.Add(sheet);
                    }
                    if (result.Count == 0) throw new 읽기오류("읽을 수 있는 시트가 없습니다.");
                    return result;
                }
            }
            catch (InvalidDataException)
            {
                throw new 읽기오류("엑셀 파일이 손상돼 열 수 없습니다.");
            }
            catch (XmlException)
            {
                throw new 읽기오류("엑셀 파일 안의 내용이 올바르지 않습니다.");
            }
        }

        static XmlDocument Part(ZipArchive zip, string path)
        {
            ZipArchiveEntry e = zip.GetEntry(path);
            if (e == null)
            {
                // 대소문자만 다른 이름으로 저장하는 프로그램이 있다.
                foreach (ZipArchiveEntry x in zip.Entries)
                    if (string.Equals(x.FullName, path, StringComparison.OrdinalIgnoreCase)) { e = x; break; }
            }
            if (e == null) return null;
            if (e.Length > 풀린최대) throw new 읽기오류("엑셀 파일 안의 내용이 너무 큽니다.");

            var doc = new XmlDocument();
            doc.XmlResolver = null;
            var settings = new XmlReaderSettings();
            settings.DtdProcessing = DtdProcessing.Prohibit;
            settings.XmlResolver = null;
            settings.MaxCharactersInDocument = 풀린최대;
            using (Stream s = e.Open())
            using (XmlReader r = XmlReader.Create(s, settings))
                doc.Load(r);
            return doc;
        }

        static List<string> SharedStrings(ZipArchive zip)
        {
            var list = new List<string>();
            XmlDocument doc = Part(zip, "xl/sharedStrings.xml");
            if (doc == null) return list;
            foreach (XmlElement si in doc.GetElementsByTagName("si", NsMain))
                list.Add(Runs(si));
            return list;
        }

        /// <summary>글자 조각(서식이 섞인 글자)을 이어 붙인다. 읽는 법 표시(rPh)는 뺀다.</summary>
        static string Runs(XmlElement parent)
        {
            var sb = new StringBuilder();
            foreach (XmlElement t in parent.GetElementsByTagName("t", NsMain))
            {
                if (t.ParentNode != null && t.ParentNode.LocalName == "rPh") continue;
                sb.Append(t.InnerText);
            }
            return sb.ToString();
        }

        static Sheet SheetOf(XmlDocument doc, List<string> shared)
        {
            var sheet = new Sheet();
            foreach (XmlElement row in doc.GetElementsByTagName("row", NsMain))
            {
                var cells = new List<string>();
                int next = 0;
                foreach (XmlNode n in row.ChildNodes)
                {
                    var c = n as XmlElement;
                    if (c == null || c.LocalName != "c") continue;
                    int col = 열번호(c.GetAttribute("r"));
                    if (col < 0) col = next;
                    next = col + 1;
                    if (col >= 최대열) continue;
                    while (cells.Count < col) cells.Add("");
                    cells.Add(값(c, shared));
                }

                // 빈 행도 자리를 지킨다 — 행 번호가 사용자 안내에 쓰인다.
                int rowNo;
                if (int.TryParse(row.GetAttribute("r"), NumberStyles.Integer, CultureInfo.InvariantCulture, out rowNo))
                    while (sheet.행.Count < rowNo - 1) 행추가(sheet, new List<string>());
                행추가(sheet, cells);
            }
            return sheet;
        }

        static string 값(XmlElement c, List<string> shared)
        {
            string t = c.GetAttribute("t");
            XmlElement v = null;
            foreach (XmlNode n in c.ChildNodes)
                if (n.LocalName == "v") { v = (XmlElement)n; break; }

            if (t == "inlineStr")
            {
                foreach (XmlNode n in c.ChildNodes)
                    if (n.LocalName == "is") return Runs((XmlElement)n);
                return "";
            }
            if (v == null) return "";
            string raw = v.InnerText;
            if (t == "s")
            {
                int i;
                return int.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out i) && i >= 0 && i < shared.Count
                    ? shared[i] : "";
            }
            if (t == "b") return raw == "1" ? "TRUE" : "FALSE";
            return raw;
        }

        /// <summary>"AB12" → 27 (0부터). 형식이 틀리면 -1.</summary>
        public static int 열번호(string cellRef)
        {
            if (string.IsNullOrEmpty(cellRef)) return -1;
            int n = 0, i = 0;
            for (; i < cellRef.Length; i++)
            {
                char ch = char.ToUpperInvariant(cellRef[i]);
                if (ch < 'A' || ch > 'Z') break;
                n = n * 26 + (ch - 'A' + 1);
                if (n > 16384) return -1;
            }
            return i == 0 ? -1 : n - 1;
        }
    }
}

File: node-server/lib/sheetreader.js
"use strict";
// src/SheetReader.cs 를 옮긴 것. ERP 엑셀(.xlsx)·CSV 를 글자 표로 읽는다.
// Node 에 들어 있는 zlib 만 쓴다 (추가 라이브러리 없음). XML 은 가볍게 훑어 값을 뽑는다.
const zlib = require("node:zlib");
const path = require("node:path");

const 최대크기 = 20 * 1024 * 1024;
const 최대행 = 20000;
const 최대열 = 200;

class 읽기오류 extends Error {}

function read(data, fileName) {
  if (!data || data.length === 0) throw new 읽기오류("빈 파일입니다.");
  if (data.length > 최대크기) throw new 읽기오류("20MB 보다 큰 파일은 읽지 않습니다.");

  if (data.length >= 4 && data[0] === 0x50 && data[1] === 0x4b && data[2] === 0x03 && data[3] === 0x04)
    return xlsx(data);

  if (data.length >= 4 && data[0] === 0xd0 && data[1] === 0xcf && data[2] === 0x11 && data[3] === 0xe0)
    throw new 읽기오류("암호가 걸렸거나 문서보안(DRM)이 적용된 파일, 또는 옛 엑셀(.xls) 파일이라 읽을 수 없습니다. " +
      "ERP 에서 CSV 로 받거나, 엑셀에서 암호·보안을 푼 뒤 .xlsx 로 저장해 올려 주세요.");

  const ext = (path.extname(fileName || "") || "").toLowerCase();
  if (ext === ".xlsx" || ext === ".xlsm")
    throw new 읽기오류("엑셀 파일 형식이 아닙니다. 파일이 손상됐거나 확장자만 바뀐 파일입니다.");

  const sheet = csv(text(data));
  sheet.이름 = path.basename(fileName || "", path.extname(fileName || ""));
  return [sheet];
}

// ── CSV ──
function text(data) {
  if (data.length >= 3 && data[0] === 0xef && data[1] === 0xbb && data[2] === 0xbf)
    return data.toString("utf8", 3);
  // UTF-8 로 읽어 보고 U+FFFD(깨짐)가 생기면 CP949 로 다시 읽는다. ERP CSV 는 대개 CP949.
  const u = data.toString("utf8");
  if (u.includes("\uFFFD")) {
    try {
      const dec = new TextDecoder("euc-kr", { fatal: false });
      return dec.decode(data);
    } catch { return u; }
  }
  return u;
}

function csv(t) {
  const sheet = { 이름: "", 행: [] };
  if (!t) return sheet;
  const eol = t.search(/[\r\n]/);
  const first = eol < 0 ? t : t.slice(0, eol);
  const sep = first.indexOf(",") < 0 && first.indexOf("\t") >= 0 ? "\t" : ",";

  let row = [], cell = "", quoted = false;
  for (let i = 0; i < t.length; i++) {
    const ch = t[i];
    if (quoted) {
      if (ch === '"') {
        if (i + 1 < t.length && t[i + 1] === '"') { cell += '"'; i++; }
        else quoted = false;
      } else cell += ch;
      continue;
    }
    if (ch === '"' && cell.length === 0) { quoted = true; continue; }
    if (ch === sep) { row.push(cell); cell = ""; continue; }
    if (ch === "\r" || ch === "\n") {
      if (ch === "\r" && i + 1 < t.length && t[i + 1] === "\n") i++;
      row.push(cell); cell = "";
      행추가(sheet, row); row = [];
      continue;
    }
    cell += ch;
  }
  if (cell.length > 0 || row.length > 0) { row.push(cell); 행추가(sheet, row); }
  return sheet;
}

function 행추가(sheet, row) {
  if (sheet.행.length >= 최대행) throw new 읽기오류("행이 너무 많습니다 (" + 최대행 + "행까지).");
  if (row.length > 최대열) row = row.slice(0, 최대열);
  sheet.행.push(row.slice());
}

// ── XLSX ──
function xlsx(data) {
  let entries;
  try { entries = unzip(data); }
  catch { throw new 읽기오류("엑셀 파일이 손상돼 열 수 없습니다."); }

  const get = (name) => {
    if (entries.has(name)) return entries.get(name);
    const lc = name.toLowerCase();
    for (const [k, v] of entries) if (k.toLowerCase() === lc) return v;
    return null;
  };

  const wb = get("xl/workbook.xml");
  if (!wb) throw new 읽기오류("엑셀 통합 문서 정보(workbook.xml)가 없습니다.");
  const wbXml = wb.toString("utf8");

  const rels = get("xl/_rels/workbook.xml.rels");
  const targets = new Map();
  if (rels) {
    const rx = /<Relationship\b([^>]*)\/?>/g;
    let m;
    while ((m = rx.exec(rels.toString("utf8")))) {
      const id = attr(m[1], "Id");
      let t = attr(m[1], "Target");
      if (!id || !t) continue;
      t = t.startsWith("/") ? t.replace(/^\/+/, "") : "xl/" + t;
      targets.set(id, t);
    }
  }

  const shared = sharedStrings(get("xl/sharedStrings.xml"));
  const result = [];
  const sx = /<sheet\b([^>]*)\/?>/g;
  let m;
  while ((m = sx.exec(wbXml))) {
    const rid = attr(m[1], "r:id");
    const name = unxml(attr(m[1], "name") || "");
    const p = targets.get(rid);
    if (!p) continue;
    const doc = get(p);
    if (!doc) continue;
    const sheet = sheetOf(doc.toString("utf8"), shared);
    sheet.이름 = name;
    result.push(sheet);
  }
  if (result.length === 0) throw new 읽기오류("읽을 수 있는 시트가 없습니다.");
  return result;
}

function attr(s, name) {
  const esc = name.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const m = new RegExp('(?:^|\\s)' + esc + '\\s*=\\s*"([^"]*)"').exec(s);
  return m ? m[1] : null;
}

function unxml(s) {
  return s.replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'").replace(/&#(\d+);/g, (_, n) => String.fromCodePoint(+n))
    .replace(/&#x([0-9a-fA-F]+);/g, (_, n) => String.fromCodePoint(parseInt(n, 16)))
    .replace(/&amp;/g, "&");
}

// <si> 안의 <t> 를 이어 붙인다. 읽는 법 표시(rPh)는 뺀다.
function sharedStrings(buf) {
  const list = [];
  if (!buf) return list;
  const xml = buf.toString("utf8");
  const sx = /<si\b[^>]*>([\s\S]*?)<\/si>/g;
  let m;
  while ((m = sx.exec(xml))) list.push(runs(m[1]));
  return list;
}

function runs(inner) {
  // rPh 블록은 통째로 뺀 뒤 <t>…</t> 만 모은다.
  const noRph = inner.replace(/<rPh\b[\s\S]*?<\/rPh>/g, "");
  let out = "";
  const tx = /<t\b[^>]*>([\s\S]*?)<\/t>|<t\b[^>]*\/>/g;
  let m;
  while ((m = tx.exec(noRph))) out += m[1] != null ? unxml(m[1]) : "";
  return out;
}

function sheetOf(xml, shared) {
  const sheet = { 이름: "", 행: [] };
  const rowx = /<row\b([^>]*)>([\s\S]*?)<\/row>|<row\b([^>]*)\/>/g;
  let rm;
  while ((rm = rowx.exec(xml))) {
    const rowAttr = rm[1] != null ? rm[1] : rm[3];
    const inner = rm[2] != null ? rm[2] : "";
    const cells = [];
    let next = 0;
    const cx = /<c\b([^>]*)>([\s\S]*?)<\/c>|<c\b([^>]*)\/>/g;
    let cm;
    while ((cm = cx.exec(inner))) {
      const cAttr = cm[1] != null ? cm[1] : cm[3];
      const cInner = cm[2] != null ? cm[2] : "";
      let col = 열번호(attr(cAttr, "r"));
      if (col < 0) col = next;
      next = col + 1;
      if (col >= 최대열) continue;
      while (cells.length < col) cells.push("");
      cells.push(값(cAttr, cInner, shared));
    }
    const rowNo = parseInt(attr(rowAttr, "r"), 10);
    if (Number.isInteger(rowNo)) while (sheet.행.length < rowNo - 1) 행추가(sheet, []);
    행추가(sheet, cells);
  }
  return sheet;
}

function 값(cAttr, cInner, shared) {
  const t = attr(cAttr, "t");
  if (t === "inlineStr") {
    const m = /<is\b[^>]*>([\s\S]*?)<\/is>/.exec(cInner);
    return m ? runs(m[1]) : "";
  }
  const vm = /<v\b[^>]*>([\s\S]*?)<\/v>/.exec(cInner);
  if (!vm) return "";
  const raw = unxml(vm[1]);
  if (t === "s") {
    const i = parseInt(raw, 10);
    return Number.isInteger(i) && i >= 0 && i < shared.length ? shared[i] : "";
  }
  if (t === "b") return raw === "1" ? "TRUE" : "FALSE";
  return raw;
}

function 열번호(cellRef) {
  if (!cellRef) return -1;
  let n = 0, i = 0;
  for (; i < cellRef.length; i++) {
    const ch = cellRef[i].toUpperCase();
    if (ch < "A" || ch > "Z") break;
    n = n * 26 + (ch.charCodeAt(0) - 64);
    if (n > 16384) return -1;
  }
  return i === 0 ? -1 : n - 1;
}

// ── 최소 ZIP 리더 (중앙 디렉터리 기준) ──
function unzip(buf) {
  const map = new Map();
  // End Of Central Directory 를 뒤에서 찾는다.
  let eocd = -1;
  for (let i = buf.length - 22; i >= 0 && i >= buf.length - 22 - 65536; i--) {
    if (buf.readUInt32LE(i) === 0x06054b50) { eocd = i; break; }
  }
  if (eocd < 0) throw new Error("ZIP EOCD 없음");
  const count = buf.readUInt16LE(eocd + 10);
  let off = buf.readUInt32LE(eocd + 16);

  for (let n = 0; n < count; n++) {
    if (buf.readUInt32LE(off) !== 0x02014b50) break;
    const method = buf.readUInt16LE(off + 10);
    const compSize = buf.readUInt32LE(off + 20);
    const nameLen = buf.readUInt16LE(off + 28);
    const extraLen = buf.readUInt16LE(off + 30);
    const commentLen = buf.readUInt16LE(off + 32);
    const localOff = buf.readUInt32LE(off + 42);
    const name = buf.toString("utf8", off + 46, off + 46 + nameLen);

    // 로컬 헤더에서 실제 데이터 시작 위치를 구한다 (extra 길이가 다를 수 있다).
    const lhNameLen = buf.readUInt16LE(localOff + 26);
    const lhExtraLen = buf.readUInt16LE(localOff + 28);
    const dataStart = localOff + 30 + lhNameLen + lhExtraLen;
    const comp = buf.subarray(dataStart, dataStart + compSize);

    if (method === 0) map.set(name, Buffer.from(comp));
    else if (method === 8) map.set(name, zlib.inflateRawSync(comp));
    // 다른 압축 방식은 건너뛴다 (xlsx 는 0/8 만 쓴다).

    off += 46 + nameLen + extraLen + commentLen;
  }
  return map;
}

module.exports = { read, csv, 읽기오류, 최대크기, 열번호 };
